### ISSUE TITLE
Cache / APCu issue

### ISSUE THREAD

**Issue by ptdev** [bcosca/fatfree#724]
There's an issue with Cache::reset() and APCu.
Previously APCu used a different key name than old APC on the values returned by apc_cache_info(). It used "key" instead of "info". This was correctly addressed by F3, but the more recent versions of APCu seem to have had that issue corrected to maintain compatibility with the old APC (see https://github.com/krakjoe/apcu/pull/42).
So, when calling Cache::reset() now on these systems we get an "Undefined index: key" error.
I've corrected the issue on my project by setting $key back to 'info' on base.php at around line 2112. Curiously, I also had to change 'mtime' to 'modification_time' at around line 2116 as it seems the more recent APCu also uses this instead of 'mtime'.
This works for my project and server setup (PHP 5.5.18 and APCu 4.0.7) but i'm posting the bug here as you'll probably want to make something more cross compatible.
Cheers.

**Comment by ikkez** [comment]
The APCu develpoer somehow messed it up (https://github.com/krakjoe/apcu/commit/4f47d6dea51cf00ab13e623a6dd91c3d96c0fcc4#diff-9476a04c2256e02b8f75fc66db3beb5eL1510) and it looks uncertain it it'll ever be fixed, so I just fixed it here.

**Comment by KOTRET** [comment]
:-1: although it seems it won't be fixed in APCu, i'm against introducing fixes for unstable bugged 3rdParty modules. My recommendation here: revert the APCu version.

**Comment by ikkez** [comment]
Hm. Well you are right and I don't like to fix bugged 3rd party php modules too. But I think this is something different than the similar issue about the bugged dblib driver, where you could easiely switch two 2 alternative drivers. 
I think there is currently no APCu version which is fully backward compatible with APC in the `apc_cache_info` function. First commit for making them compatible was made by @bcosca himself (https://github.com/ikkez/fatfree/commit/9cb5a05875ee530ad3c4ceb0952abeea087bd4ce). I could imagine that some clients do not have a change to select which APC / APCu version they want to be running on their hosting, so most times the developer have to deal with what is given. The alternative to not use APCu doesn't look like an option. @ptdev Do you have any more information about the current status about this issue on APCu? Is it fixed, will it be fixed or is APCu already dead?
When I ask google for "apc_cache_info apcu", it just shows a lot of issues.
Maybe we just simplify the code here to something like:

```php
case 'apc':
  $key='key';
  $mtkey='mtime';
case 'apcu':
  $key='info';
  $mtkey='modification_time';
```
Will this be enough, at least for the lasted APCu version, @ptdev ?

**Comment by KOTRET** [comment]
you cannot just `case` the different apc versions, because older versions of apcu still work properly with a key named `info` (this will result in f3 only supporting the latest (bugged) version of apcu. The approach you tried is correct, i just don't feel comfortable with the problem itself ;)

**Comment by xfra35** [comment]
I think @ikkez's patch is the best we can do at the moment.

So let's keep it and review it later when APCu has been patched (which could happen in a long time..).

**Comment by ptdev** [comment]
Hey all, although I still haven't tested @ikkez 's patch, a quick look at it seems to be a suitable fix for the issue, at least until a fully backwards compatible version of APCu is released and we can (i guess) define a minimum version requirement for it.

That being said, it does appear that the APCu developer is aware of the issue and tried to fix it before but still committed a bugged version. There are several issues on their tracker as well as commit comments that address this problem so I'm pretty confident that the next version would have the issue finally fixed. The question now is, when will it be fixed if ever. Also, if fixed, how long until a fixed version gets into PECL which would also be kind of a necessity so most *nix package managers can upgrade it and release it to the masses.

I'm guessing as more and more developers migrate to 5.5+, the need for a suitable "user cache" will increase since the "old" APC got discontinued in favor of the new Zend opcode caching thing, leaving no suitable or backwards compatible user caching mechanism. At least none that is as used and "battle tested" as APC which is still a critical performance need for many applications.

**Comment by KOTRET** [comment]
> it does appear that the APCu developer is aware of the issue

sure, but taking a look into commit history the project seems to be inactive...

**Comment by Rayne** [comment]
:thumbsup: for patch https://github.com/bcosca/fatfree-core/commit/cd55cb41a21c05395749c9eb9462971cd0278471 and the comments of @xfra35 (https://github.com/bcosca/fatfree/issues/724#issuecomment-104628756) and @ikkez (https://github.com/bcosca/fatfree/issues/724#issuecomment-104584934).

**Comment by bcosca** [comment]
Is this still an issue with the apcu-bc module?

**Comment by ptdev** [comment]
I'm using the latest version available on debian stable which is still 4.0.7 and it seems to be working fine. I assume that's because the framework is detecting the key name (info or key) as per this issue and respective fixes.

From apcu's changelog, it appears as the later versions have had the issue fixed (I'm guessing that's what the "fix inconsistent member names for entries in userland" on 4.0.8 means)

As for apcu-bc, it seems to be php7 only and requires apcu 5.1.2
